subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java

```diff
@@ -102,68 +106,76 @@ private void deleteDistributions(Collection<File> dirs) {
     }
 
     private Multimap<GradleVersion, File> determineChecksumDirsByVersion() {
-        Multimap<GradleVersion, File> result = ArrayListMultimap.create();
+        Multimap<GradleVersion, File> checksumDirsByVersion = ArrayListMultimap.create();
         for (File dir : listDirs(distsDir)) {
             for (File checksumDir : listDirs(dir)) {
-                try {
-                    GradleVersion gradleVersion = determineGradleVersionFromBuildReceipt(checksumDir);
-                    result.put(gradleVersion, checksumDir);
-                } catch (Exception e) {
-                    // TODO see https://github.com/gradle/gradle-private/issues/1379
-                    // LOGGER.debug("Could not determine Gradle version for {}", checksumDir, e);
+                Result<GradleVersion> result = determineGradleVersionFromBuildReceipt(checksumDir);
+                if (result.hasValue()) {
+                    checksumDirsByVersion.put(result.getValue(), checksumDir);
+                } else {
+                    LOGGER.debug("Could not determine Gradle version for {}: {}", checksumDir, result.getError());
                 }
             }
         }
-        return result;
+        return checksumDirsByVersion;
     }
 
-    private GradleVersion determineGradleVersionFromBuildReceipt(File checksumDir) throws Exception {
+    private Result<GradleVersion> determineGradleVersionFromBuildReceipt(File checksumDir) {
         List<File> subDirs = listDirs(checksumDir);
-        Preconditions.checkArgument(subDirs.size() == 1, "A Gradle distribution must contain exactly one subdirectory: %s", subDirs);
+        if (subDirs.size() != 1) {
+            return error("A Gradle distribution must contain exactly one subdirectory: {0}", subDirs);
```

Comment:
If `Result` is only used to avoid printing stack traces, maybe we can do like this:

```
private Optional< GradleVersion> error(String messageTemplate, Object ...arguments) {
     LOGGER.debug(messageTemplate, arguments);
     return Optional.empty();
}
```